kv/LevelDBStore, FileStore, MonDBStore: simpler code for single-key fetches

branch-predictor opened the pull request:
Often there's a need to fetch value for a specified, single key. Publish
a new method in LevelDB that takes key and prefix and puts just value at
specified address, without the need for interim keysets or resultsets.
Make use of that method in MonitorDBStore and LevelDB (not all cases, only
the most often called ones).

Signed-off-by: Piotr Dałek piotr.dalek@ts.fujitsu.com

Commit:
kv/LevelDBStore: simpler code for single-key fetches

Often there's a need to fetch value for a specified, single key. Publish
a new method in LevelDB that takes key and prefix and puts just value at
specified address, without the need for interim keysets or resultsets.

Signed-off-by: Piotr Dałek <piotr.dalek@ts.fujitsu.com>

liewegas commented on src/kv/LevelDBStore.cc:
```diff
@@ -218,6 +218,22 @@ int LevelDBStore::get(
+  KeyValueDB::Iterator it = get_iterator(prefix);
+  it->lower_bound(key);
+  if (it->valid() && it->key() == key) {
+      *value = it->value();
```
i think we need to else return -ENOENT.  can probably mostly cut&paste from here:

https://github.com/liewegas/ceph/blob/wip-kv/src/kv/RocksDBStore.cc#L296

liewegas commented on src/mon/MonitorDBStore.h:
```diff
@@ -507,14 +507,12 @@ class MonitorDBStore
+    if (outbl.length() == 0) 
       return -ENOENT;
-    bl.append(out[key]);
+    bl.append(outbl);
```
yes

liewegas commented on src/os/DBObjectMap.cc:
```diff
@@ -1089,17 +1089,16 @@ DBObjectMap::Header DBObjectMap::_lookup_map_header(
   Header ret(header, RemoveOnDelete(this));
-  bufferlist::iterator iter = out.begin()->second.begin();
+  bufferlist::iterator iter = out.begin();
+
```
yes

let's break these 2 other hunks into separate patches.  they work already since there is the generic single-key get wrapper in KeyValueDB

Commit:
os/DBObjectMap: use single-key LevelDB::get method

Make FileStore use single-key ::get() method for some performance
increase in k/v-heavy workloads.

Signed-off-by: Piotr Dałek <piotr.dalek@ts.fujitsu.com>

Commit:
mon/MonitorDBStore: use single-key LevelDB::get() method

Make MonitorDBStore use single-key ::get() method for some performance
increase in heavy K/V workloads.

Signed-off-by: Piotr Dałek <piotr.dalek@ts.fujitsu.com>

branch-predictor commented:
@liewegas done.
